app/src/main/java/com/example/androidtraining/MainViewModel.kt

```diff
@@ -0,0 +1,72 @@
+package com.example.androidtraining
+
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.example.androidtraining.models.LoginResponse
+import com.example.androidtraining.models.Product
+import com.example.androidtraining.models.UserInfo
+import com.example.androidtraining.networking.TokenProvider
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@HiltViewModel
+class MainViewModel @Inject constructor(private val firstRepository: FirstRepository) :
+    ViewModel() {
+    @Inject
+    lateinit var tokenProvider: TokenProvider
+    private val _response = MutableStateFlow<Result<LoginResponse>?>(null)
+    val response: StateFlow<Result<LoginResponse>?> = _response
+
+    private val _successfulLogin = MutableStateFlow<Boolean?>(null)
+    val successfulLogin: StateFlow<Boolean?> = _successfulLogin
+
+    private val _isLoading = MutableStateFlow(false)
+    var isLoading: StateFlow<Boolean> = _isLoading
+
+    fun login(userInfo: UserInfo) {
```

Comment:
Note: Structure your code in accordance to types. A general guide is properties > custom constructors (if any) > functions. Having a function in the middle of property declarations is reducing code clarity.